quantity/core.py

```diff
@@ -0,0 +1,329 @@
+                    exp = exp.__float__()
+                except Exception:
+                    return NotImplemented
+            else:
+                if exp.imag == 0:
+                    exp = exp.real
+
+        return self.__class__(self.value ** exp, self.unit ** exp)
+
+    def _operate(self, other, op, units_helper):
+        if not isinstance(other, QUANTITY_OR_NUMBER) and (
+                not has_array_namespace(other)
+        ):
+            # HACK: unit should take care of this!
+            if not isinstance(other, u.UnitBase):
+                return NotImplemented
+
+            try:
+                unit = getattr(self.unit, op.replace("__i", "__"))(other)
+            except Exception:
+                return NotImplemented
+            else:
+                return self.__class__(self.value, unit)
+
+        other_value, other_unit = get_value_and_unit(other)
+
+        self_value = self.value
+        (conv0, conv1), unit = units_helper(self.unit, other_unit)
+
+        if conv0 is not None:
+            self_value = conv0(self_value)
+        if conv1 is not None:
+            other_value = conv1(other_value)
+        try:
+            value = getattr(self_value, op)(other_value)
+        except AttributeError:
+            return NotImplemented
+
+        return self.__class__(value, unit)
+
+    __array_function__ = None
+
+    def __array_ufunc__(self, function, method, *inputs, **kwargs):
+        # Used for our namespace.
+        if (where := kwargs.get("where")) is not None and isinstance(where, Quantity):
+            return NotImplemented
+
+        converters, unit = converters_and_unit(function, method, *inputs)
+        out = kwargs.get("out")
+        if out is not None:
+            # If pre-allocated output is used, check it is suitable.
+            # This also returns array view, to ensure we don't loop back.
+            out_values = self._check_output(out, unit if len(out) > 1 else [unit], function=function)
+            # Ensure output argument remains a tuple.
+            kwargs["out"] = out_values
+            if len(out) == 1:
+                out = out[0]
+
+        if method == "reduce" and "initial" in kwargs and unit is not None:
+            # Special-case for initial argument for reductions like
+            # np.add.reduce.  This should be converted to the output unit as
+            # well, which is typically the same as the input unit (but can
+            # in principle be different: unitless for np.equal, radian
+            # for np.arctan2, though those are not necessarily useful!)
+            kwargs["initial"] = self._to_own_unit(kwargs["initial"], unit=unit)
+
+        input_values = [get_value_and_unit(in_)[0] for in_ in inputs]
+        input_values = [v if conv is None else conv(v) for (v, conv) in zip(input_values, converters, strict=True)]
+        try:
+            xp = self.value.__array_namespace__()
+        except AttributeError:
+            xp = np
```

Comment:
what about defining a ``__value_array_namespace__`` that does this?

```python
    def __value_array_namespace__(self):
        try:
            xp = self.value.__array_namespace__()
        except AttributeError:
            xp = np
        return xp
```

Reply:
IDK. If we are strict that `Quantity` only works with Array API (which I think we should be) then we don't need this at all.